Support for PLAIN BB-Code and CODE=rich

asccc opened the pull request:
Fixes #29
Fixes #30
Fixes #32
Fixes #37 

I also added phpunit and wrote some tests as mentioned in #31
We can add phpunit to our phan action, too.

asccc committed:
had to reset my branch

asccc committed:
remove unused files in parser-directory after reset

jr-cologne commented on app/CodeFormatterApp.php:
```diff
@@ -52,39 +53,64 @@ private function parseInput(): iterable
     /**
      * formats a token based on its type and language.
```
```suggestion
     * formats a node based on its type and language.
```

jr-cologne commented on tests/Parser/ParserTest.php:
```diff
@@ -0,0 +1,125 @@
+    {
+        $parser = new Parser();
+        return iterator_to_array($parser->parseText($input));
+    }
```
I feel like it would be nice to extract this into a helper trait.

jr-cologne commented on tests/Parser/ParserTest.php:
```diff
@@ -0,0 +1,125 @@
+
+    public function testBracketsInsideTextAreHandledCorrectly()
+    {
+        $nodes = $this->parseTextToArray('hello[[[[[');
+        $this->assertCount(2, $nodes);
+        $this->assertInstanceof(TextNode::class, $nodes[0]);
+        $this->assertEquals('hello', $nodes[0]->getBody());
+        $this->assertInstanceof(TextNode::class, $nodes[1]);
+        $this->assertEquals('[[[[[', $nodes[1]->getBody());
```
I wonder why you chose to interpret this as two different text nodes rather than just a single one.

asccc replied:
The current algorithm uses `[` to split text-nodes, because each `[` could start a new bbcode. I'm not satisfied with this either, that's why the pr is still a draft :)

asccc replied:
Fixed in https://github.com/dev-community-de/code-formatter/pull/36/commits/0f57d541b345c4b6c82e676eb185019c2e36cf40

jr-cologne commented on tests/Parser/ParserTest.php:
```diff
@@ -0,0 +1,125 @@
+
+    public function testNestedCodeBbcodesBehaveTheSameAsInXenforo()
+    {
+        $nodes = $this->parseTextToArray('[code][code]test[/code][/code]');
```
If possible, I would parse this as an element/code node consisting of a body `[code]test[/code]`.
In the forum, we could then even support syntax highlighting of BB code.

asccc replied:
I'm afraid this is not possible for the same reason as `</script>` cannot occur inside `<script>`. 

For example:

`[code=css].foo::before { content: '[code]test[/code]'; }[/code]`

There is no way to parse this code-element correctly without interpreting the code-content. 
But in order to interpret the code-content, we have to parse the code-element first ... do you see the problem?

![grafik](https://user-images.githubusercontent.com/42143765/87355061-fb7c1700-c55f-11ea-9e41-f76c74eceabb.png)

We could allow nested [code] elements, but XenForo does not support such nesting and would break anyway.

jr-cologne replied:
Ah, see your point. Sure thing.
Thanks for the explanation.

asccc replied:
It would be possible now after the CODE=rich update (the necessary code to do this is there).
But XenForo doesn't support it ... is there a way to escape [/code] inside a code-element?

jr-cologne commented on app/Parser/Parser.php:
```diff
@@ -4,33 +4,173 @@
+            $position = stripos($textData, $needle, $start);
+            if (false === $position) {
+                // edge case:
+                // no closing tag found... tread the rest of the input
```
```suggestion
                // no closing tag found... treat the rest of the input
```

jr-cologne reviewed (commented):
Yep, adding phpunit to our GitHub actions would be cool.

Thank you for fixing/addressing these bugs/issues as well as your additional improvements of the code in general.

I have already provided some feedback, even though you are probably still working on it.
I will go ahead and fix your mistakes in the comments and extract the helper methods for the tests into a trait.

jr-cologne committed:
Fix typos/mistakes in code comments

jr-cologne committed:
Properly set up phpunit testsuite and extract helper methods to trait

jr-cologne committed:
Add phpunit to phan github action

asccc committed:
simplify the parser and make sure text-nodes are buffered

asccc commented:
I had to remove the parser-trait, because currently there is no autoloader in place to load classes inside the tests-folder.

asccc committed:
stop parsing after unclosed bbcode

there is no reason the check the text-buffer after a unclosed bbcode,
because the unclosed bbcode consumes the rest of the input anyway.

asccc commented:
Argh, i missed the change in composer.json ... autoloading in tests/ was possible ☹️

jr-cologne commented:
Yep, I actually set everything up for phpunit.

asccc committed:
fix wording in test

asccc committed:
fix wording in test #2

asccc committed:
add comment about nested bbcodes in test

asccc committed:
added parser-helpers trait again

asccc committed:
used more descriptive variable-names in parser

asccc committed:
used more descriptive variable-names in parser

asccc committed:
Merge branch 'parser' of github.com:asccc/code-formatter into parser

asccc commented:
I think we're good to go now 😃

asccc committed:
hide phan progress-bar, because github cannot render it

asccc commented:
I decided to implement CODE=rich (as mentioned in #37) in this pr, too, because this requires another rewrite of the parser.
It wouldn't make sense to merge this right now.

jr-cologne commented:
Alright, take your time. All looks very good to me. 😃

asccc committed:
(wip) added support for nested code-blocks

asccc committed:
some fixes and adjusted node-export logic

asccc committed:
tests added

asccc committed:
fix code=rich export issue

asccc commented on app/CodeFormatterApp.php:
```diff
@@ -32,59 +33,96 @@ public function __construct(Parser $parser = null)
+            foreach ($nodeBody as $subNode) {
+                $buffer .= $this->formatNode($subNode);
+            }
+            return $this->exportNode($node, $buffer);
```
I'm not sure if this is an issue, but CODE=rich skips the formatting part. 
We can change this, but i currently see no reason to format "rich" code (nested code-nodes get formatted as before).

jr-cologne replied:
No, this is totally fine. 😃

asccc commented:
Should work now 😃

asccc committed:
small adjustments in architecture + saner api

asccc commented:
Somehow my latest push is not showing up here?

asccc commented:
Well, anyway @jr-cologne i'm not 100% satisfied with the api. What do you think?
You can see the latest changes in my fork: https://github.com/asccc/code-formatter/commit/40c28034cdac188f0b1e0fdbe4bec4618ef8fc76

asccc committed:
test if pr is broken

asccc committed:
yep, pr is broken

jr-cologne commented:
Sorry, I have kinda lost track of the changes you have applied here. 😄 
I will try to get an overview over your code as well as the API in general in order to answer your question.
Just give me a moment to take a closer look at it.

jr-cologne commented on app/Parser/NodeList.php:
```diff
@@ -0,0 +1,65 @@
+        return $this->nodes;
+    }
+
+    /**
+     * returns the node-list as iterable list.
+     *
+     * (phan has issues with returning an array as iterable)
+     *
+     * @phan-suppress PhanParamSignatureMismatchInternal
+     *
+     * @return Node[]
+     */
+    public function getIterator(): iterable
+    {
+        return $this->nodes;
+    }
```
Are you sure phan complains without any good reason here?
I'm getting the following error:
```php
Fatal error: Uncaught Exception: Objects returned by DevCommunityDE\CodeFormatter\Parser\NodeList::getIterator() must be traversable or implement interface Iterator in /srv/app/app/CodeFormatterApp.php:76
```
I'm not sure, but could it be possible that we have to construct and return a specific iterator like [`ArrayIterator`](https://www.php.net/manual/de/arrayiterator.construct.php)?

```suggestion
    /**
     * returns the node-list as iterable list.
     *
     * @return Node[]
     */
    public function getIterator(): iterable
    {
        return new ArrayIterator($this->nodes);
    }
```

jr-cologne reviewed (commented):
Ok, to me, the API and such looks perfectly fine.
What do you exactly not like yet?

Skipping the formatter if the language is "text" would probably speed things up a little bit as the formatter wouldn't have to go through all potential formatters to finally find anything which formats plain text. So feel free to make this change.

asccc committed:
fix IteratorAggregate usage

asccc commented:
I absolutely don't like the fact that ElemNode's body can be a NodeList or simply a string.
This requires assertions all over the place and makes the tests type unsafe:

![grafik](https://user-images.githubusercontent.com/42143765/87613031-6cf8c880-c70c-11ea-93e2-0662c296e923.png)

(All the red highlights are failed type-checks).

asccc committed:
make tests type-safe

asccc committed:
added more tests and fixed some associated issues

asccc committed:
fix typo in file-path

asccc commented:
"Besser wirds erstmal nicht" i guess 😆

asccc committed:
fixed some typos and added more tests

asccc committed:
use expectException in tests

jr-cologne committed:
Merge branch 'parser' of https://github.com/asccc/code-formatter into pr/36

jr-cologne commented:
Oops, I accidentally pushed something. I hope I've not broken anything. 😅 
Looks like to merged your parser branch...

jr-cologne commented:
Ok, besides one error I get when formatting with prettier, everything looks pretty good. 👍 

Here's the output:
```
Fatal error: Uncaught DevCommunityDE\CodeFormatter\Exceptions\Exception: code formatting failed! stderr: Error: EACCES: permission denied, mkdir '/var/www/.npm' command not found: prettier
```

Test input:
```
Hallo
[CODE=css].test { padding: 10px; }[/CODE]
test
```

This might just be an issue on my end, btw.
If it works for you, feel free to merge.

Before you merge and I forget it: Thank you so much for this huge amount of work you put into this PR! 🥇 🥳 🙌
